pkg/webhooks/validation_test.go

```diff
@@ -0,0 +1,543 @@
+	"github.com/kyverno/kyverno/pkg/engine/context"
+	"github.com/kyverno/kyverno/pkg/engine/response"
+	"github.com/kyverno/kyverno/pkg/engine/utils"
+	"gotest.tools/assert"
+)
+
+func TestValidate_failure_action_overrides(t *testing.T) {
+
+	testcases := []struct {
+		rawPolicy   []byte
+		rawResource []byte
+		blocked     bool
+	}{
+		{
+			rawPolicy: []byte(`
+				{
+					"apiVersion": "kyverno.io/v1",
+					"kind": "ClusterPolicy",
+					"metadata": {
+					   "name": "check-label-app"
+					},
+					"spec": {
+					   "validationFailureAction": "audit",
+					   "validationFailureActionOverrides": 
+							[
+								{
+									"action": "enforce",
+									"namespaces": [
+										"default"
+									]
+								},
+								{
+									"action": "audit",
+									"namespaces": [
+										"test"
+									]
+								}
+							],
+					   "rules": [
+						  {
+							 "name": "check-label-app",
+							 "match": {
+								"resources": {
+								   "kinds": [
+									  "Pod"
+								   ]
+								}
+							 },
+							 "validate": {
+								"message": "The label 'app' is required.",
+								"pattern": {
+									"metadata": {
+										"labels": {
+											"app": "?*"
+										}
+									}
+								}
+							}
+						  }
+					   ]
+					}
+				 }
+			`),
+			rawResource: []byte(`
+				{
+					"apiVersion": "v1",
+					"kind": "Pod",
+					"metadata": {
+					   "name": "test-pod",
+					   "namespace": "default"
+					},
+					"spec": {
+					   "containers": [
+						  {
+							 "name": "nginx",
+							 "image": "nginx:latest"
+						  }
+					   ]
+					}
+				 }
+			`),
+			blocked: true,
+		},
+		{
+			rawPolicy: []byte(`
+				{
+					"apiVersion": "kyverno.io/v1",
+					"kind": "ClusterPolicy",
+					"metadata": {
+					   "name": "check-label-app"
+					},
+					"spec": {
+					   "validationFailureAction": "audit",
+					   "validationFailureActionOverrides": 
+							[
+								{
+									"action": "enforce",
+									"namespaces": [
+										"default"
+									]
+								},
+								{
+									"action": "audit",
+									"namespaces": [
+										"test"
+									]
+								}
+							],
+					   "rules": [
+						  {
+							 "name": "check-label-app",
+							 "match": {
+								"resources": {
+								   "kinds": [
+									  "Pod"
+								   ]
+								}
+							 },
+							 "validate": {
+								"message": "The label 'app' is required.",
+								"pattern": {
+									"metadata": {
+										"labels": {
+											"app": "?*"
+										}
+									}
+								}
+							}
+						  }
+					   ]
+					}
+				 }
+			`),
+			rawResource: []byte(`
+				{
+					"apiVersion": "v1",
+					"kind": "Pod",
+					"metadata": {
+					   "name": "test-pod",
+					   "labels": {
+						   "app": "my-app"
+					   }
+					},
+					"spec": {
+					   "containers": [
+						  {
+							 "name": "nginx",
+							 "image": "nginx:latest"
+						  }
+					   ]
+					}
+				 }
+			`),
+			blocked: false,
```

Comment:
The pod is in default namespace, shouldn't it be blocked?

Reply:
Ah my bad! I thought it only tests `validationFailureActionOverrides`, didn't notice the rule pattern.